File: leetcode_oct_daily_challenge_2021/longestCommonSubsequence/iterativeDP.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
      
        dp = [[0] * len(text2) for i in range(len(text1))]
        
        for i in range(len(text1)):
            for j in range(len(text2)):
                if text1[i] == text2[j]:
                    if i == 0 or j == 0:
                        dp[i][j] = 1
                    else:
                        dp[i][j] = 1 + dp[i - 1][j - 1]
                else:
                    dp[i][j] = max(dp[i - 1][j] , dp[i][j - 1])
                    
        return dp[-1][-1]

File: leetcode_oct_daily_challenge_2021/longestCommonSubsequence/iterativeDP.java
class Solution {
  public int longestCommonSubsequence(String str1, String str2) {
      int[][] dp = new int[str1.length() + 1][str2.length() + 1];
      
      for(int i = str1.length() - 1; i >= 0; i--){
          for(int j = str2.length() - 1; j >= 0; j--){
              if(str1.charAt(i) == str2.charAt(j))
                  dp[i][j] = dp[i+1][j+1] + 1;
              else
                  dp[i][j] = Math.max(dp[i+1][j], dp[i][j+1]);
          }
      }
      return dp[0][0];
  }
}